### PR TITLE
feat: Add fallback to Bunny URLs if main URLs are not accessible

### PR DESCRIPTION
**Pull request by TanmayDhobale**
### PR Fixes: #776 
This PR adds a new feature to the `ContentRenderer` component that checks if the main video URLs from the database are accessible or not. If any of the main URLs are not accessible, it falls back to using the Bunny URLs instead.

Changes:

1. Added a new helper function `isUrlAccessible` that makes a HEAD request to a given URL and returns a boolean indicating whether the URL is accessible or not.

2. Modified the `getMetadata` function:
   - If the `user.bunnyProxyEnabled` flag is true, it returns the Bunny URLs directly.
   - If `bunnyProxyEnabled` is false, it creates an object `mainUrls` containing the main URLs from the database.
   - It then uses `Promise.all` and the `isUrlAccessible` function to check if all the main URLs are accessible.
   - If all main URLs are accessible, it returns `mainUrls`.
   - If any of the main URLs is not accessible, it returns the Bunny URLs instead.

This change ensures that if the main video URLs are not accessible for any reason, the application falls back to using the Bunny URLs, providing a more reliable experience for users.

### Checklist before requesting a review
- [x] I have performed a self-review of my code
- [x] I assure there is no similar/duplicate pull request regarding same issue

### CODE REVIEW

**Comment by TanmayDhobale** [comment]
@hkirat  This approach is right ? like rn it makes HEAD requests to all the main URLs, which could potentially cause performance issues if there are many URLs to check Instead of checking the accessibility of all main URLs (1080p, 720p, and 360p) by making individual HEAD requests for each URL we can  only checks the accessibility of the highest quality video URL (1080p).

**Comment by siinghd** [comment]
> @hkirat This approach is right ? like rn it makes HEAD requests to all the main URLs, which could potentially cause performance issues if there are many URLs to check Instead of checking the accessibility of all main URLs (1080p, 720p, and 360p) by making individual HEAD requests for each URL we can only checks the accessibility of the highest quality video URL (1080p).

This should work fine, but can you implement a fallback mechanism that attempts to check other urls only if necessary?

what i mean is: 

1- check the highest quality url first, if available `return mainurls` if not 
2- check for 720 and 360, if 720 available `return mainurls`, if not return check for 360, if 360 available `return mainurls` else return bunnyurls

**Comment by TanmayDhobale** [comment]
> > @hkirat This approach is right ? like rn it makes HEAD requests to all the main URLs, which could potentially cause performance issues if there are many URLs to check Instead of checking the accessibility of all main URLs (1080p, 720p, and 360p) by making individual HEAD requests for each URL we can only checks the accessibility of the highest quality video URL (1080p).
> 
> This should work fine, but can you implement a fallback mechanism that attempts to check other urls only if necessary?
> 
> what i mean is: 
> 
> 1- check the highest quality url first, if available `return mainurls` if not 
> 2- check for 720 and 360, if 720 available `return mainurls`, if not check for 360, if 360 available `return mainurls` else return bunnyurls

Yeah sure, let me do that. ( We got a new maintainer before GTA 6)

**Comment by TanmayDhobale** [comment]
@siinghd done

File: src/components/admin/ContentRenderer.tsx [hunk 1]
```diff
@@ -86,24 +85,52 @@ export const getMetadata = async (contentId: number) => {
...
+      720: metadata[`video_720p_mp4_${userId}`],
+      360: metadata[`video_360p_mp4_${userId}`],
+      subtitles: metadata['subtitles'],
+      slides: metadata['slides'],
+      segments: metadata['segments'],
+    };
+  }
+
+  // If the highest quality URL is not accessible, check the 720p URL
+  const mediumQualityUrl = metadata[`video_720p_mp4_${userId}`];
```
Comment (siinghd) [inline src/components/admin/ContentRenderer.tsx:105]: can we improve the code?

```
const mainUrls = {
    1080: metadata[`video_1080p_mp4_${userId}`],
    720: metadata[`video_720p_mp4_${userId}`],
    360: metadata[`video_360p_mp4_${userId}`],
    subtitles: metadata['subtitles'],
    slides: metadata['slides'],
    segments: metadata['segments'],
    thumbnails: metadata['thumbnail_mosiac_url'],
  };

  const bunnyUrls = {
    1080: bunnyUrl(metadata[`video_1080p_mp4_${userId}`]),
    720: bunnyUrl(metadata[`video_720p_mp4_${userId}`]),
    360: bunnyUrl(metadata[`video_360p_mp4_${userId}`]),
    subtitles: metadata['subtitles'],
    slides: metadata['slides'],
    segments: metadata['segments'],
    thumbnails: metadata['thumbnail_mosiac_url'],
  };

  const isHighestQualityUrlAccessible = await isUrlAccessible(mainUrls['1080']);

  if (isHighestQualityUrlAccessible) {
    return mainUrls;
  }

  const otherQualities = ['720', '360'];
  for (const quality of otherQualities) {
    const urlKey = `${quality}`;
    const isAccessible = await isUrlAccessible(mainUrls[urlKey]);
    if (isAccessible) {
      return mainUrls;
    }
  }

  // If none of the main URLs are accessible, return Bunny URLs
  return bunnyUrls;
```


something like this?

Comment (TanmayDhobale) [reply src/components/admin/ContentRenderer.tsx]: > can we improve the code?
> 
> ```
> const mainUrls = {
>     1080: metadata[`video_1080p_mp4_${userId}`],
>     720: metadata[`video_720p_mp4_${userId}`],
>     360: metadata[`video_360p_mp4_${userId}`],
>     subtitles: metadata['subtitles'],
>     slides: metadata['slides'],
>     segments: metadata['segments'],
>     thumbnails: metadata['thumbnail_mosiac_url'],
>   };
> 
>   const bunnyUrls = {
>     1080: bunnyUrl(metadata[`video_1080p_mp4_${userId}`]),
>     720: bunnyUrl(metadata[`video_720p_mp4_${userId}`]),
>     360: bunnyUrl(metadata[`video_360p_mp4_${userId}`]),
>     subtitles: metadata['subtitles'],
>     slides: metadata['slides'],
>     segments: metadata['segments'],
>     thumbnails: metadata['thumbnail_mosiac_url'],
>   };
> 
>   const isHighestQualityUrlAccessible = await isUrlAccessible(mainUrls['1080']);
> 
>   if (isHighestQualityUrlAccessible) {
>     return mainUrls;
>   }
> 
>   const otherQualities = ['720', '360'];
>   for (const quality of otherQualities) {
>     const urlKey = `${quality}`;
>     const isAccessible = await isUrlAccessible(mainUrls[urlKey]);
>     if (isAccessible) {
>       return mainUrls;
>     }
>   }
> 
>   // If none of the main URLs are accessible, return Bunny URLs
>   return bunnyUrls;
> ```
> 
> something like this?

sure , thats cleaner lol

File: src/components/admin/ContentRenderer.tsx [hunk 2]
```diff
@@ -81,65 +81,47 @@ export const getMetadata = async (contentId: number) => {
...
+    360: metadata[`video_360p_mp4_${userId}`],
+    subtitles: metadata['subtitles'],
+    slides: metadata['slides'],
+    segments: metadata['segments'],
+    thumbnails: metadata['thumbnail_mosiac_url'],
+  };
 
-  // If none of the main URLs are accessible, return Bunny URLs
-  return {
+  const bunnyUrls = {
```
Comment (siinghd) [inline src/components/admin/ContentRenderer.tsx:98]: this can be moved up right? at line `75` right? so at line 77 we can return bunnyUrls ?

Comment (siinghd) [reply src/components/admin/ContentRenderer.tsx:98]: jsut fix that and ill merge it when i get the access

Comment (TanmayDhobale) [reply src/components/admin/ContentRenderer.tsx]: u dont have access ? sad

**Comment by siinghd** [comment]
@TanmayDhobale 

still the same thing

here:

```
if (user.bunnyProxyEnabled) {
    return {
      1080: bunnyUrl(metadata[`video_1080p_mp4_${userId}`]),
      720: bunnyUrl(metadata[`video_720p_mp4_${userId}`]),
      360: bunnyUrl(metadata[`video_360p_mp4_${userId}`]),
      subtitles: metadata['subtitles'],
      slides: metadata['slides'],
      segments: metadata['segments'],
      thumbnails: metadata['thumbnail_mosiac_url'],
    };
  }
  
  this is duplicate
```
  
  {
      1080: bunnyUrl(metadata[`video_1080p_mp4_${userId}`]),
      720: bunnyUrl(metadata[`video_720p_mp4_${userId}`]),
      360: bunnyUrl(metadata[`video_360p_mp4_${userId}`]),
      subtitles: metadata['subtitles'],
      slides: metadata['slides'],
      segments: metadata['segments'],
      thumbnails: metadata['thumbnail_mosiac_url'],
    };
```

```

**Comment by siinghd** [comment]
Move const bunnyUrls = {

above 
```

if (user.bunnyProxyEnabled) {
    return {
      1080: bunnyUrl(metadata[`video_1080p_mp4_${userId}`]),
      720: bunnyUrl(metadata[`video_720p_mp4_${userId}`]),
      360: bunnyUrl(metadata[`video_360p_mp4_${userId}`]),
      subtitles: metadata['subtitles'],
      slides: metadata['slides'],
      segments: metadata['segments'],
      thumbnails: metadata['thumbnail_mosiac_url'],
    };
  }
  
  and do 
  
  if (user.bunnyProxyEnabled) {
    return bunnyUrls
  }
```

**Comment by TanmayDhobale** [comment]
now 😭 @siinghd

**Comment by TanmayDhobale** [comment]
@siinghd

File: src/components/admin/ContentRenderer.tsx [hunk 3]
```diff
@@ -64,52 +72,48 @@ export const getMetadata = async (contentId: number) => {
...
+
+  const isHighestQualityUrlAccessible = await isUrlAccessible(mainUrls['1080']);
+
+  if (isHighestQualityUrlAccessible) {
+    return mainUrls;
+  }
+
+  const otherQualities = ['720', '360'];
+  for (const quality of otherQualities) {
+    const urlKey = `${quality}`;
```
Comment (hkirat) [inline src/components/admin/ContentRenderer.tsx:108]: urlKey = `${quality}` ?
Why not urlKey = quality?
Even better do u need urlKey?

Comment (TanmayDhobale) [reply src/components/admin/ContentRenderer.tsx]: there was no need for the urlKey variable,  I have removed the urlKey variable altogether and directly used the quality value to access the property from the mainUrls object.

**Comment by siinghd** [comment]
Merging Thank you, will submit this pr, if it will satisfy bounty requirements, you will get it

**Comment by TanmayDhobale** [comment]
> Merging Thank you, will submit this pr, if it will satisfy bounty requirements, you will get it

sure 👍

**Pull request merged**